packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,87 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:collection';
+
+import 'asset_bundle.dart';
+
+/// Base class for resolver components that transform an input into part of
+/// a resolved asset's key.
+abstract class Qualifier {
+  String resolve();
+}
+
+/// Qualifier that knows how to find the nearest match among a list of
+/// candidates, based on a mapping to candidates from their normative values.
+/// The mapping should be sorted, e.g. by passing a [SplayTreeMap].
+///
+/// The resolution is done via linear search based on the assumption that the
+/// number of candidates is short.
```

Comment:
I looked into this. SplayTreeMap has a pair of find functions that give you the nearest element above or below the input key value.  Of course, that would mean forcing clients to use SplayTreeMap, when a sorted LinkedHashMap is the natural choice for material/icon.dart.

If you want to define a literal to use with this class, I think you can't do better than linear, because either you build a SplayTreeMap (n*logn) or you do a binary search on a non-random-access data structure (again n*logn).